diff --git a/absl/container/internal/btree.h b/absl/container/internal/btree.h
index 01f4e749..3ff8e798 100644
--- a/absl/container/internal/btree.h
+++ b/absl/container/internal/btree.h
@@ -75,16 +75,6 @@ namespace absl {
 ABSL_NAMESPACE_BEGIN
 namespace container_internal {
 
-#ifdef ABSL_BTREE_ENABLE_GENERATIONS
-#error ABSL_BTREE_ENABLE_GENERATIONS cannot be directly set
-#elif defined(ABSL_HAVE_ADDRESS_SANITIZER) || \
-    defined(ABSL_HAVE_MEMORY_SANITIZER)
-// When compiled in sanitizer mode, we add generation integers to the nodes and
-// iterators. When iterators are used, we validate that the container has not
-// been mutated since the iterator was constructed.
-#define ABSL_BTREE_ENABLE_GENERATIONS
-#endif
-
 template <typename Compare, typename T, typename U>
 using compare_result_t = absl::result_of_t<const Compare(const T &, const U &)>;
 
