diff --git a/src/bthread/task_group.cpp b/src/bthread/task_group.cpp
index 6f5a4abd..e3aef91c 100644
--- a/src/bthread/task_group.cpp
+++ b/src/bthread/task_group.cpp
@@ -248,7 +248,7 @@ int TaskGroup::init(size_t runqueue_capacity) {
     return 0;
 }

-#if defined(__linux__) && defined(__aarch64__) && defined(__clang__)
+#if defined(__clang__)
     __attribute__((optnone))
 #endif
 void TaskGroup::task_runner(intptr_t skip_remained) {
@@ -570,7 +570,7 @@ void TaskGroup::sched(TaskGroup** pg) {
     sched_to(pg, next_tid);
 }

-#if defined(__linux__) && defined(__aarch64__) && defined(__clang__)
+#if defined(__clang__)
     __attribute__((optnone))
 #endif
 void TaskGroup::sched_to(TaskGroup** pg, TaskMeta* next_meta) {
