diff --git a/tools/rpc_press/rpc_press_impl.cpp b/tools/rpc_press/rpc_press_impl.cpp
index 8a873eff..d1ff2a1e 100644
--- a/tools/rpc_press/rpc_press_impl.cpp
+++ b/tools/rpc_press/rpc_press_impl.cpp
@@ -222,7 +222,7 @@ void RpcPress::sync_client() {
     int64_t last_expected_time = butil::monotonic_time_ns();
     const int64_t interval = (int64_t) (1000000000L / req_rate);
     // the max tolerant delay between end_time and expected_time. 10ms or 10 intervals
-    int64_t max_tolerant_delay = std::max(10000000L, 10 * interval);    
+    int64_t max_tolerant_delay = std::max<int64_t>(10000000L, 10 * interval);
     while (!_stop) {
         brpc::Controller* cntl = new brpc::Controller;
         msg_index = (msg_index + _options.test_thread_num) % _msgs.size();
