diff --git a/src/bvar/variable.cpp b/src/bvar/variable.cpp
index 1f9da3a..f57aaa6 100644
--- a/src/bvar/variable.cpp
+++ b/src/bvar/variable.cpp
@@ -81,11 +81,7 @@ struct VarMapWithLock : public VarMap {
 
     VarMapWithLock() {
         CHECK_EQ(0, init(1024, 80));
-        pthread_mutexattr_t attr;
-        pthread_mutexattr_init(&attr);
-        pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_RECURSIVE);
-        pthread_mutex_init(&mutex, &attr);
-        pthread_mutexattr_destroy(&attr);
+        pthread_mutex_init(&mutex, NULL);
     }
 };
 
