--- a/configure	2022-11-19 23:27:17.324633792 +0800
+++ b/configure	2022-11-19 23:26:58.612602601 +0800
@@ -17099,7 +17099,7 @@
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lgssapi_krb5 ${GSSAPIBASE_LIBS} -lgssapi_krb5 -lkrb5 -lk5crypto -lcom_err ${LIB_SOCKET} $LIBS"
+LIBS="-lgssapi_krb5 ${GSSAPIBASE_LIBS} -lgssapi_krb5 -lkrb5 -lk5crypto -lcom_err -lkrb5support -ldl -lresolv ${LIB_SOCKET} $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -17306,8 +17306,8 @@
   fi
 
   if test "$gss_impl" = "mit"; then
-    GSSAPIBASE_LIBS="$GSSAPIBASE_LIBS -lgssapi_krb5 -lkrb5 -lk5crypto -lcom_err"
-    GSSAPIBASE_STATIC_LIBS="$GSSAPIBASE_LIBS $gssapi_dir/libgssapi_krb5.a $gssapi_dir/libkrb5.a $gssapi_dir/libk5crypto.a $gssapi_dir/libcom_err.a"
+    GSSAPIBASE_LIBS="$GSSAPIBASE_LIBS -lgssapi_krb5 -lkrb5 -lk5crypto -lcom_err -lkrb5support -ldl -lresolv"
+    GSSAPIBASE_STATIC_LIBS="$GSSAPIBASE_LIBS $gssapi_dir/libgssapi_krb5.a $gssapi_dir/libkrb5.a $gssapi_dir/libk5crypto.a $gssapi_dir/libcom_err.a $gssapi_dir/libkrb5support.a"
   elif test "$gss_impl" = "heimdal"; then
     CPPFLAGS="$CPPFLAGS"
     GSSAPIBASE_LIBS="$GSSAPIBASE_LIBS -lgssapi -lkrb5 -lasn1 -lroken ${LIB_CRYPT} ${LIB_DES} -lcom_err"

--- a/lib/dlopen.c	2022-11-21 02:42:16.590169904 +0800
+++ b/lib/dlopen.c	2022-11-21 02:42:51.202224311 +0800
@@ -56,10 +56,8 @@
 #include <sasl.h>
 #include "saslint.h"
 
-#ifndef PIC
 #include <saslplug.h>
 #include "staticopen.h"
-#endif
 
 #ifdef DO_DLOPEN
 #if HAVE_DIRENT_H
@@ -414,11 +412,9 @@
     DIR *dp;
     struct dirent *dir;
 #endif
-#ifndef PIC
     add_plugin_t *add_plugin;
     _sasl_plug_type type;
     _sasl_plug_rec *p;
-#endif
 
     if (! entrypoints
 	|| ! getpath_cb
@@ -429,7 +425,6 @@
 	|| ! verifyfile_cb->proc)
 	return SASL_BADPARAM;
 
-#ifndef PIC
     /* do all the static plugins first */
 
     for(cur_ep = entrypoints; cur_ep->entryname; cur_ep++) {
@@ -456,7 +451,6 @@
 	    	result = add_plugin(p->name, p->plug);
 	}
     }
-#endif /* !PIC */
 
 /* only do the following if:
  * 
