--- a/include/flatbuffers/stl_emulation.h	2021-05-11 02:45:16.000000000 +0800
+++ b/include/flatbuffers/stl_emulation.h	2025-01-06 22:43:47.538175771 +0800
@@ -625,7 +625,7 @@
  private:
   // This is a naive implementation with 'count_' member even if (Extent != dynamic_extent).
   pointer const data_;
-  const size_type count_;
+  size_type count_;
 };
 
  #if !defined(FLATBUFFERS_SPAN_MINIMAL)
