--- glog-0.3.3/src/vlog_is_on.cc	2012-01-12 16:40:58.000000000 +0800
+++ glog-0.3.3/src/vlog_is_on.cc	2015-03-22 16:48:50.759037426 +0800
@@ -165,7 +165,8 @@
   int result = FLAGS_v;
   int const pattern_len = strlen(module_pattern);
   bool found = false;
-  MutexLock l(&vmodule_lock);  // protect whole read-modify-write
+  // MutexLock l(&vmodule_lock);  // protect whole read-modify-write
+  vmodule_lock.Lock();  // protect whole read-modify-write
   for (const VModuleInfo* info = vmodule_list;
        info != NULL; info = info->next) {
     if (info->module_pattern == module_pattern) {
@@ -189,6 +189,8 @@
     info->next = vmodule_list;
     vmodule_list = info;
   }
+  vmodule_lock.Unlock();
+
   RAW_VLOG(1, "Set VLOG level for \"%s\" to %d", module_pattern, log_level);
   return result;
 }
