--- glog-0.3.3/src/utilities.h.orig 2017-08-31 15:38:59.050406303 +0800
+++ glog-0.3.3/src/utilities.h  2017-08-31 15:40:02.015350365 +0800
@@ -101,17 +101,17 @@
 // correctly when GetStackTrace() is called with max_depth == 0.
 // Some code may do that.
 
-#if defined(HAVE_LIB_UNWIND)
-# define STACKTRACE_H "stacktrace_libunwind-inl.h"
-#elif !defined(NO_FRAME_POINTER)
-# if defined(__i386__) && __GNUC__ >= 2
-#  define STACKTRACE_H "stacktrace_x86-inl.h"
-# elif defined(__x86_64__) && __GNUC__ >= 2 && HAVE_UNWIND_H
-#  define STACKTRACE_H "stacktrace_x86_64-inl.h"
-# elif (defined(__ppc__) || defined(__PPC__)) && __GNUC__ >= 2
-#  define STACKTRACE_H "stacktrace_powerpc-inl.h"
-# endif
-#endif
+// #if defined(HAVE_LIB_UNWIND)
+// # define STACKTRACE_H "stacktrace_libunwind-inl.h"
+// #elif !defined(NO_FRAME_POINTER)
+// # if defined(__i386__) && __GNUC__ >= 2
+// #  define STACKTRACE_H "stacktrace_x86-inl.h"
+// # elif defined(__x86_64__) && __GNUC__ >= 2 && HAVE_UNWIND_H
+// #  define STACKTRACE_H "stacktrace_x86_64-inl.h"
+// # elif (defined(__ppc__) || defined(__PPC__)) && __GNUC__ >= 2
+// #  define STACKTRACE_H "stacktrace_powerpc-inl.h"
+// # endif
+// #endif
 
 #if !defined(STACKTRACE_H) && defined(HAVE_EXECINFO_H)
 # define STACKTRACE_H "stacktrace_generic-inl.h"
