--- ./src/core/lib/iomgr/tcp_posix_old.cc	2025-07-19 02:19:55
+++ ./src/core/lib/iomgr/tcp_posix.cc	2025-07-19 02:19:44
@@ -22,6 +22,8 @@
 
 #include "src/core/lib/iomgr/exec_ctx.h"
 #include "src/core/lib/iomgr/port.h"
+
+#include "absl/strings/str_cat.h"
 
 #ifdef GRPC_POSIX_SOCKET_TCP
 
--- ./src/core/lib/iomgr/event_engine_shims/tcp_client_old.cc	2025-07-19 02:24:05
+++ ./src/core/lib/iomgr/event_engine_shims/tcp_client.cc	2025-07-19 02:16:24
@@ -17,6 +17,7 @@
 
 #include "absl/status/status.h"
 #include "absl/strings/string_view.h"
+#include "absl/strings/str_cat.h"
 
 #include <grpc/event_engine/event_engine.h>
 #include <grpc/support/time.h>
