--- a/src/core/lib/promise/detail/basic_seq.h	2025-01-06 22:41:37.857651534 +0800
+++ b/src/core/lib/promise/detail/basic_seq.h	2025-01-06 22:41:42.308703853 +0800
@@ -471,7 +471,7 @@
           cur_ = next;
           state_.~State();
           Construct(&state_,
-                    Traits::template CallSeqFactory(f_, *cur_, std::move(arg)));
+                    Traits::CallSeqFactory(f_, *cur_, std::move(arg)));
           return PollNonEmpty();
         });
   }
