--- include/jemalloc/jemalloc.sh
+++ include/jemalloc/jemalloc.sh
@@ -12,7 +12,7 @@ extern "C" {
 EOF

 for hdr in jemalloc_defs.h jemalloc_rename.h jemalloc_macros.h \
-           jemalloc_protos.h jemalloc_typedefs.h jemalloc_mangle.h ; do
+           jemalloc_protos.h jemalloc_typedefs.h ; do
   cat "${objroot}include/jemalloc/${hdr}" \
       | grep -v 'Generated from .* by configure\.' \
       | sed -e 's/ $//g'