--- a/libdivide.h	2025-01-06 22:27:33.308725176 +0800
+++ b/libdivide.h	2025-01-06 22:27:41.517821664 +0800
@@ -3017,7 +3017,7 @@
     T recover() const { return div.recover(); }
 
     bool operator==(const divider<T, ALGO> &other) const {
-        return div.denom.magic == other.denom.magic && div.denom.more == other.denom.more;
+        return div.denom.magic == other.div.denom.magic && div.denom.more == other.div.denom.more;
     }
 
     bool operator!=(const divider<T, ALGO> &other) const { return !(*this == other); }
