--- a/config.h.in	2012-05-29 01:11:28.000000000 +0800
+++ b/config.h.in	2022-05-11 18:06:02.000000000 +0800
@@ -176,7 +176,7 @@
 #undef HAVE_GSSAPI_H

 /* Define to 1 if you have the `GSS_C_NT_HOSTBASED_SERVICE' function. */
-#undef HAVE_GSS_C_NT_HOSTBASED_SERVICE
+#define HAVE_GSS_C_NT_HOSTBASED_SERVICE 1

 /* Define to 1 if you have the `gss_decapsulate_token' function. */
 #undef HAVE_GSS_DECAPSULATE_TOKEN
@@ -185,10 +185,10 @@
 #undef HAVE_GSS_ENCAPSULATE_TOKEN

 /* Define to 1 if you have the `gss_inquire_mech_for_saslname' function. */
-#undef HAVE_GSS_INQUIRE_MECH_FOR_SASLNAME
+#define HAVE_GSS_INQUIRE_MECH_FOR_SASLNAME 1

 /* Define to 1 if you have the `gss_oid_equal' function. */
-#undef HAVE_GSS_OID_EQUAL
+#define HAVE_GSS_OID_EQUAL 1

 /* Define if you have the iconv() function and it works. */
 #undef HAVE_ICONV
