--- src/rdkafka_broker.c	2021-10-12 04:15:44.000000000 +0800
+++ src/rdkafka_broker.c	2022-03-10 10:31:45.141882467 +0800
@@ -5464,7 +5464,9 @@
  */
 void rd_kafka_broker_destroy_final (rd_kafka_broker_t *rkb) {
 
-        rd_assert(thrd_is_current(rkb->rkb_thread));
+        // To avoid the error describe in https://github.com/edenhill/librdkafka/issues/3608
+        // comment this line to fix it temporarily.
+        // rd_assert(thrd_is_current(rkb->rkb_thread));
         rd_assert(TAILQ_EMPTY(&rkb->rkb_monitors));
         rd_assert(TAILQ_EMPTY(&rkb->rkb_outbufs.rkbq_bufs));
         rd_assert(TAILQ_EMPTY(&rkb->rkb_waitresps.rkbq_bufs));

--- mklove/modules/configure.base	2022-11-21 03:47:14.072094876 +0800
+++ mklove/modules/configure.base	2022-11-21 03:47:01.360075164 +0800
@@ -1716,7 +1716,7 @@
     mkl_check_begin "$cname" "$2" "no-cache" "$1 (by pkg-config)" && return $?
 
     local cflags=
-    local cmd="${PKG_CONFIG} --short-errors --cflags $libname"
+    local cmd="${PKG_CONFIG} --static --short-errors --cflags $libname"
     mkl_dbg "pkg-config check $libname for CFLAGS ($2): $cmd"
 
     cflags=$($cmd 2>&1)
@@ -1738,11 +1738,11 @@
     fi
 
     local libs=
-    cmd="${PKG_CONFIG} --short-errors --libs $libname"
+    cmd="${PKG_CONFIG} --static --short-errors --libs $libname"
     mkl_dbg "pkg-config check $libname for LIBS ($2): $cmd"
     libs=$($cmd 2>&1)
     if [[ $? != 0 ]]; then
-        mkl_dbg "${PKG_CONFIG} --libs $libname failed: $libs"
+        mkl_dbg "${PKG_CONFIG} --static --libs $libname failed: $libs"
         # Clear define name ($2): caller may have additional checks
         mkl_check_failed "$cname" "" "$3" "pkg-config --libs failed"
         return 1
