--- autoconf/config.guess    2013-07-26 09:35:43.000000000 +0800
+++ autoconf/config.guess    2019-12-24 16:00:25.940000000 +0800
@@ -864,6 +864,13 @@
     i*86:Minix:*:*)
 	echo ${UNAME_MACHINE}-pc-minix
 	exit ;;
+    aarch64:Linux:*:*)
+    echo "$UNAME_MACHINE"-unknown-linux-"$LIBC"
+    exit ;;
+    aarch64_be:Linux:*:*)
+    UNAME_MACHINE=aarch64_be
+    echo "$UNAME_MACHINE"-unknown-linux-"$LIBC"
+    exit ;;
     alpha:Linux:*:*)
     case `sed -n '/^cpu model/s/^.*: \(.*\)/\1/p' < /proc/cpuinfo` in
       EV5)   UNAME_MACHINE=alphaev5 ;;
