--- cmd-line-utils/libedit/terminal.c.orig	2017-05-24 19:02:28.440144595 -0700
+++ cmd-line-utils/libedit/terminal.c	2017-05-24 19:03:03.552146135 -0700
@@ -866,7 +866,7 @@
 terminal_set(EditLine *el, const char *term)
 {
 	int i;
-	char buf[TC_BUFSIZE];
+	// char buf[TC_BUFSIZE];
 	char *area;
 	const struct termcapstr *t;
 	sigset_t oset, nset;
@@ -876,7 +876,8 @@
 	(void) sigaddset(&nset, SIGWINCH);
 	(void) sigprocmask(SIG_BLOCK, &nset, &oset);
 
-	area = buf;
+	// area = buf;
+	area = NULL;
 
 
 	if (term == NULL)
