diff -uprN a/cmake/zlib.cmake b/cmake/zlib.cmake
--- a/cmake/zlib.cmake  2022-01-18 00:21:44.993277439 +0800
+++ b/cmake/zlib.cmake  2022-01-18 00:21:21.943260690 +0800
@@ -52,7 +52,7 @@ MACRO (MYSQL_CHECK_ZLIB_WITH_COMPRESS)
     INCLUDE(FindZLIB)
     IF(ZLIB_FOUND)
      INCLUDE(CheckFunctionExists)
-      SET(CMAKE_REQUIRED_LIBRARIES z)
+      SET(CMAKE_REQUIRED_LIBRARIES ${ZLIB_LIBRARIES})
       CHECK_FUNCTION_EXISTS(crc32 HAVE_CRC32)
       CHECK_FUNCTION_EXISTS(compressBound HAVE_COMPRESSBOUND)
       CHECK_FUNCTION_EXISTS(deflateBound HAVE_DEFLATEBOUND)
diff -uprN a/cmake/boost.cmake b/cmake/boost.cmake
--- a/cmake/boost.cmake	2021-09-28 17:56:01.656714647 +0800
+++ b/cmake/boost.cmake	2021-09-28 17:58:09.156339908 +0800
@@ -31,7 +31,7 @@
 # we assume that the correct version (see below)
 # is installed on the compile host in the standard location.
 
-SET(BOOST_PACKAGE_NAME "boost_1_59_0")
+SET(BOOST_PACKAGE_NAME "boost_1_73_0")
 SET(BOOST_TARBALL "${BOOST_PACKAGE_NAME}.tar.gz")
 SET(BOOST_DOWNLOAD_URL
   "http://sourceforge.net/projects/boost/files/boost/1.59.0/${BOOST_TARBALL}"
@@ -262,7 +262,7 @@ IF(NOT BOOST_MAJOR_VERSION EQUAL 10)
   COULD_NOT_FIND_BOOST()
 ENDIF()
 
-IF(NOT BOOST_MINOR_VERSION EQUAL 59)
+IF(NOT BOOST_MINOR_VERSION EQUAL 73)
   MESSAGE(WARNING "Boost minor version found is ${BOOST_MINOR_VERSION} "
     "we need 59"
     )
diff -uprN a/cmd-line-utils/libedit/terminal.c b/cmd-line-utils/libedit/terminal.c
--- a/cmd-line-utils/libedit/terminal.c	2021-09-28 17:56:01.661714985 +0800
+++ b/cmd-line-utils/libedit/terminal.c	2021-09-28 17:56:52.794174061 +0800
@@ -866,7 +866,6 @@ protected int
 terminal_set(EditLine *el, const char *term)
 {
 	int i;
-	char buf[TC_BUFSIZE];
 	char *area;
 	const struct termcapstr *t;
 	sigset_t oset, nset;
@@ -876,7 +875,7 @@ terminal_set(EditLine *el, const char *t
 	(void) sigaddset(&nset, SIGWINCH);
 	(void) sigprocmask(SIG_BLOCK, &nset, &oset);
 
-	area = buf;
+	area = NULL;
 
 
 	if (term == NULL)
diff -uprN a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt    2022-01-23 23:30:48.915006943 +0800
+++ b/CMakeLists.txt    2022-01-23 23:31:04.785141731 +0800
@@ -484,8 +484,6 @@ INCLUDE_DIRECTORIES(
 MYSQL_CHECK_ZLIB_WITH_COMPRESS()
 # Add bundled yassl/taocrypt or system openssl.
 MYSQL_CHECK_SSL()
-# Add system/bundled editline.
-MYSQL_CHECK_EDITLINE()
 # Add libevent
 MYSQL_CHECK_LIBEVENT()
 # Add lz4 library
