--- a/include/sqltypes.h        2022-04-19 14:37:42.584467011 +0800
+++ b/include/sqltypes.h        2022-04-19 13:56:41.486446873 +0800
@@ -82,13 +82,13 @@
  * 	by the standard linux string header files.
  */
 #ifdef SQL_WCHART_CONVERT
-typedef wchar_t             TCHAR;
+typedef wchar_t             TWCHAR;
 #else
-typedef signed short        TCHAR;
+typedef signed short        TWCHAR;
 #endif
 
 #else
-typedef char				TCHAR;
+typedef char				TWCHAR;
 #endif
 
 typedef unsigned short		WORD;
@@ -108,7 +108,7 @@
 typedef WCHAR* 	            LPWSTR;
 typedef const char*         LPCSTR;
 typedef const WCHAR*        LPCWSTR;
-typedef TCHAR*              LPTSTR;
+typedef TWCHAR*              LPTSTR;
 typedef char*               LPSTR;
 typedef DWORD*           	LPDWORD;
 
--- a/include/sqlext.h	2023-12-05 15:56:09.623865253 +0800
+++ a/include/sqlext.h	2023-12-05 15:55:28.852847283 +0800
@@ -2190,7 +2190,7 @@ void	FireVSDebugEvent(PODBC_VS_ARGS);
  * connection pooling retry times
  */

-BOOL ODBCSetTryWaitValue ( DWORD dwValue );
+int ODBCSetTryWaitValue ( DWORD dwValue );
 #ifdef __cplusplus
 DWORD ODBCGetTryWaitValue ( );
 #else

--- a/include/sqltypes.h	2023-12-05 15:56:24.372871756 +0800
+++ a/include/sqltypes.h	2023-12-05 15:17:26.558878980 +0800
@@ -68,7 +68,7 @@ extern "C" {
 #else
 #define SQL_API
 #endif
-#define	BOOL				int
+// #define	BOOL				int
 #ifndef _WINDOWS_
 typedef void*				HWND;
 #endif
