diff --git a/entrypoint.sh b/entrypoint.sh
index d3399e3..5c8ee7b 100755
--- a/entrypoint.sh
+++ b/entrypoint.sh
@@ -202,7 +202,7 @@ if ((CHECKBASHISMS_ENABLE == 1)); then
 fi
 
 if ((shellcheck_code != 0 || shfmt_code != 0)); then
-	if [ "$GITHUB_EVENT_NAME" == "pull_request" ] && ((SH_CHECKER_COMMENT == 1)); then
+	if [[ "$GITHUB_EVENT_NAME" == "pull_request" || "$GITHUB_EVENT_NAME" == "pull_request_target" ]] && ((SH_CHECKER_COMMENT == 1)); then
 		_comment_on_github "$shellcheck_error" "$shfmt_error"
 	fi
 fi
